Share expiring messages only to relays that support expiring messages

martindsq opened the pull request:
Closes #312 

When opening sockets, we are now checking if the associated relays have a metadata, if they don't or the metadata was last queried more than 3 days ago, we fetch the metadata and store the result in a new coredata class.
When publishing a expiring note to a single relay or to all relays, we first check that the selected relay supports NIP 40 or that there is at least one relay (if the user didn't select one) that supports it.
Relays that don't publish a list if supported NIPs are treated as if they didn't support any of them.
When debugging, the list of relays shows the metadata fetched from them.

martindsq commented on RelayMetadata+CoreDataClass.swift:
```diff
@@ -0,0 +1,56 @@
+        timestamp = Date.now
+    }
+
+    override public var description: String {
```
This is just used for debugging purposes

martindsq commented on Nos/Models/Relay+CoreDataClass.swift:
```diff
@@ -93,6 +94,13 @@ public class Relay: NosManagedObject {
+    class func find(supporting nip: Int, for author: Author, context: NSManagedObjectContext) async throws -> [Relay] {
+        try await context.perform {
+            let relays = try context.fetch(Relay.relays(for: author))
+            return relays.filter { $0.metadata?.supportedNIPs?.contains(nip) ?? false }
```
There isn't a way to do this filter using a predicate unless we refactor the list of supported nips as a one-to-many relationship that I find too much or a string with ints separated by something. Given that the list of relays shouldn't be too large, I think we can get away with this simple filter.

mplorentz commented on CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+- Fixed a bug where ixpred messages could be published to relays that doesn't support them
```
```suggestion
- Fixed a bug where expired messages could be published to relays that doesn't support them
```

mplorentz commented on Nos/Assets/Localization/Localized.swift:
```diff
@@ -103,6 +103,8 @@ enum Localized: String, Localizable, CaseIterable {
     case editProfile = "Edit Profile"
     case youNeedToEnterAPrivateKeyBeforePosting = "You need to enter a private key in Settings before you can publish posts."
+    case relayDoesNotSupportNIP40 = "This relay doesn't support expiring messages. Please select another one."
+    case anyRelaysSupportingNIP40 = "There aren't any relays that support expiring messages in your relay list. Please add one and retry."
```
I liked Sebastian's suggestion of "None of your relays support expiring messages", do you?
```suggestion
    case anyRelaysSupportingNIP40 = "None of your relays support expiring messages. Please add one and retry."
```

mplorentz commented on Nos/Models/Nos.xcdatamodeld/.xccurrentversion:
```diff
@@ -3,6 +3,6 @@
 <dict>
 	<key>_XCCurrentVersionName</key>
-	<string>Nos 6.xcdatamodel</string>
+	<string>Nos 7.xcdatamodel</string>
```
U

mplorentz commented on RelayMetadata+CoreDataClass.swift:
```diff
@@ -0,0 +1,56 @@
+import Foundation
+import CoreData
+
+public class RelayMetadata: NosManagedObject {
```
Given that this model has a 1-to-1 relationship with `Relay`, should we just move all the properties on `RelayMetadata` to `Relay`? It will make them easier and cheaper to access.

martindsq replied:
Done

mplorentz reviewed (commented):
This is looking very good. I left some feedback and as discussed in demo yesterday people would like to see the relay metadata cleaned up a little bit and added to the release builds.

Also I think this branch will need to be merge after `fix-spinners` since we both added version 7 of the core data model but the `fix-spinner`s one is already in production. `fix-spinners` has an open PR but in the meantime you could merge that branch into this one and set it as the target of this PR. If you need help resolving conflicts let me know, but I won't be working much next week so it might be a while.

martindsq commented on Nos/Service/RelayService.swift:
```diff
@@ -491,14 +502,67 @@ extension RelayService {
+            guard let relay = try backgroundContext.fetch(Relay.relay(by: address)).first else {
+                return false
+            }
+            guard let timestamp = relay.metadata?.timestamp else {
```
👌

martindsq commented:
@mplorentz can you test this branch again? I changed the base to fix-spinners. It's working for me in my device. But my simulators are working really weird in this one and in fix-spinners but not in main: for example, I cannot get past the first screen that creates an account on a fresh install, the button doesn't react at all.

martindsq commented on Nos/Service/RelayService.swift:
```diff
@@ -368,22 +368,33 @@ extension RelayService {
     }
-    
+
     func publish(
```
Let's better fix it there so if a bug arises we are sure where it happened.

martindsq commented on Nos/Models/Relay+CoreDataClass.swift:
```diff
@@ -93,6 +94,17 @@ public class Relay: NosManagedObject {
+    ) async throws -> [Relay] {
+        try await context.perform {
+            let relays = try context.fetch(Relay.relays(for: author))
+            return relays.filter { $0.metadata?.supportedNIPs?.contains(nipNumber) ?? false }
```
I commented above that I found creating a separate model too much engeneering. I created a ticket for doing that in the future https://github.com/planetary-social/nos/issues/400

mplorentz commented on Nos.xcodeproj/project.pbxproj:
```diff
@@ -2198,15 +2186,18 @@
 				C9DEBFD6298941000078B43A /* Nos.xcdatamodel */,
 			);
-			currentVersion = 5B503F632A291D480098805A /* Nos 7.xcdatamodel */;
+			currentVersion = 5BFF66AF2A4B55FC00AA79DD /* Nos 10.xcdatamodel */;
```
Do we need Nos 9.xcdatamodel? I think `fix-spinners` added Nos 8, so this PR should add Nos 9? Am I missing something?

martindsq replied:
No, we don't need it. fix-spinners added 7 and 8 as well, so I though it was ok to maintain intermediate models. The number 9 was the datamodel as I developed it before collapsing RelayMetadata properties to Relay directly, maybe we can delete 9 and rename 10 later.

martindsq replied:
I removed previous versions

mplorentz commented on Nos/Models/Nos.xcdatamodeld/Nos 7.xcdatamodel/contents:
I don't think we should be modifying version 7 in this PR? Anyway I deleted it on `fix-spinners`.

martindsq replied:
👌

mplorentz reviewed (commented):
I made one more tweak on `fix-spinners` which you should merge in. I deleted all the xcdatamodel versions before 8 so we can finally get rid of the "does not have an inverse" warnings we were getting in the old models. I was able to do this because I dropped the database on `fix-spinners`. Sorry to cause more conflicts but I really hate those warnings 😓 

The code changes here look fine to me. While testing I seem to have only gotten metadata for one relay, and I'm pretty sure relay.damus.io and e.nos.lol have metadata. I didn't see any obvious reason why in the logs.

martindsq commented:
@mplorentz I can retrieve the metadata of those two relays without any problem. I will check it out later, but please keep an eye on it, maybe its just a matter of retrying

mplorentz reviewed (approved):
This is looking good! My app fetched the metadata for all the relays this time. One thing I would still like to see addressed in this PR are these build warnings:
![Screenshot 2023-07-06 at 10 12 01 AM](https://github.com/planetary-social/nos/assets/1165004/b078631b-9c9b-4063-bc0f-058ca6885212)

I also would like to see the relay metadata displayed on a separate screen that is shown when tapping the relay row as we discussed in the demo. That doesn't need to be done in this PR, and I'm happy to do it myself if you prefer. I'm pretty sure I could whip it out in 20 minutes or so.

martindsq commented:
I fixed the openSockets and supportedNIPS warnings and added a separate screen for the metadata:

![Captura de pantalla 2023-07-06 a la(s) 16 05 28](https://github.com/planetary-social/nos/assets/1703242/279b069f-f8a6-476a-80ee-d89714b0f1ca)
